[charts-pro] Add a `borderRadius` property to `FunnelChart` (#17660)
mui/mui-x: docs/data/charts/funnel/funnel.md

{{"demo": "FunnelGap.js"}}

### Border radius

The border radius of the sections can be customized by the `borderRadius` property.
It accepts a number that represents the radius in pixels.

> - The `bump` curve interpolation will not respect the border radius.

Reviewer: Why?

Author: Couldn't get the math right. Too complex. 

Like, the bump is a quadratic bezier curve like in the left below, with border radius, in the other curves I do something like the right. 

But doing the curve correctly between the two secondary points without weird bumps proved difficult to me. I spent a few hours on it, and then decided it wasn't worth it 😆  

![image](https://github.com/user-attachments/assets/37e78109-5c84-45e6-b570-ce3e489e5324)

Reviewer: Sounds good. Should we try to explain to what extent the linear curve respects the border radius? 

From the point of view of someone reading the docs, "the linear curve respects the border radius to some extent" doesn't provide a lot of clarity. Can we be more specific about what are the problems of using a linear curve with the border radius?

Author: it is a bit hard to explain to the user. We could try to more closely follow the border radius, but the shape looks odd quite fast, this is 10px radius

<img width="380" alt="Screenshot 2025-05-06 at 13 16 38" src="https://github.com/user-attachments/assets/d6448c70-3e62-44a0-a680-1b4b46fbe342" />

Reviewer: I see. Since there's the demo below, I'm ok with keeping it as is, as the user can see for themselves.

> - The `linear` curve will respect the border radius to some extent due to the angle of the sections.

Reviewer: ```suggestion
- The `linear` curve respects the border radius to some extent due to the angle of the sections.
```

Can we be more specific of the problems? To which extent isn't the radius respected, or why isn't it respected?

- The `step` curve will respect the border radius.

> You can play with this in the [curve interpolation example](/x/react-charts/funnel/#curve-interpolation) above.

Reviewer: ```suggestion
You can experience how the border radius interacts with the `curve` prop in the [curve interpolation example](/x/react-charts/funnel/#curve-interpolation).
```

> The `borderRadius` property will also behave differently depending if the `gap` property is bigger than 0 or not.

Reviewer: Should we be more specific? What do we mean by differently in this case?

Author: added more context

{{"demo": "FunnelGap.js"}}

### Border radius

The border radius of the sections can be customized by the `borderRadius` property.
It accepts a number that represents the radius in pixels.

- The `bump` curve interpolation will not respect the border radius.
- The `linear` curve respects the border radius to some extent due to the angle of the sections.
- The `step` curve will respect the border radius.

To understand how the border radius interacts with the `curve` prop, see the [curve interpolation example](/x/react-charts/funnel/#curve-interpolation) above.

> The `borderRadius` property will also behave differently depending if the `gap` property is bigger than 0 or not.

Reviewer: ```suggestion
The `borderRadius` property will also behave differently depending on whether the `gap` property is greater than 0.
```

- If the `gap` is 0, the border radius will be applied to the corners of the sections that are not connected to another section.

> - If the `gap` is bigger than 0, the border radius will be applied to all the corners of the sections.

Reviewer: ```suggestion
- If the `gap` is greater than 0, the border radius will be applied to all the corners of the sections.
```
